src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,469 @@
+        // now see that there are enough or too few entries, we stop
+        // removing our entry.
+
+        if (aNumEntries <= aDesiredNumEntries)
+        {
+            SetState(kAdded);
+        }
+        break;
+    }
+}
+
+void Publisher::HandleTimer(Timer &aTimer)
+{
+    aTimer.Get<Publisher>().HandleTimer();
+}
+
+void Publisher::HandleTimer(void)
+{
+    if (mState == kAdding)
+    {
+        Add();
+    }
+    else if (mState == kRemoving)
+    {
+        Remove(/* aNextState */ kToAdd);
+    }
+}
+
+void Publisher::Add(void)
+{
+    // Adds the service entry to the network data.
+
+    switch (mType)
+    {
+    case kTypeAnycast:
+        SuccessOrExit(Get<Service::Manager>().Add<Service::DnsSrpAnycast>(
+            Service::DnsSrpAnycast::ServiceData(static_cast<uint8_t>(mPortOrSeqNumber))));
+        break;
+
+    case kTypeUnicast:
+        SuccessOrExit(Get<Service::Manager>().Add<Service::DnsSrpUnicast>(
+            Service::DnsSrpUnicast::ServiceData(mAddress, mPortOrSeqNumber)));
+        break;
+
+    case kTypeUincastMeshLocalEid:
+        SuccessOrExit(Get<Service::Manager>().Add<Service::DnsSrpUnicast>(
+            Service::DnsSrpUnicast::ServerData(mAddress, mPortOrSeqNumber)));
+        break;
+    }
+
+    SetState(kAdded);
+
+    Get<Notifier>().HandleServerDataUpdated();
+
+exit:
+    return;
+}
+
+void Publisher::Remove(State aNextState, bool aRegisterWithLeader)
+{
+    // Removes the service entry from network data (if it was added).
+
+    VerifyOrExit((mState == kAdded) || (mState == kRemoving));
+
+    switch (mType)
+    {
+    case kTypeAnycast:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpAnycast>(
+            Service::DnsSrpAnycast::ServiceData(static_cast<uint8_t>(mPortOrSeqNumber))));
+        break;
+
+    case kTypeUnicast:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpUnicast>(
+            Service::DnsSrpUnicast::ServiceData(mAddress, mPortOrSeqNumber)));
+        break;
+
+    case kTypeUincastMeshLocalEid:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpUnicast>());
+        break;
+    }
+
+    if (aRegisterWithLeader)
+    {
+        Get<Notifier>().HandleServerDataUpdated();
+    }
+
+    otLogInfoNetData("Publisher: Removed DNS/SRP service from network data");
+
+exit:
+    SetState(aNextState);
+}
+
+void Publisher::HandleNotifierEvents(Events aEvents)
+{
+    if ((mType == kTypeUincastMeshLocalEid) && aEvents.Contains(kEventThreadMeshLocalAddrChanged))
+    {
+        mAddress = Get<Mle::Mle>().GetMeshLocal64();
+
+        if (mState == kAdded)
+        {
+            // If the entry is already added, we need to update it
+            // so we remove it and add it back immediately with
+            // the new mesh-local address.
+
+            Remove(/* aNextState */ kAdding);
+            Add();
+        }
+    }
+
+    if (aEvents.ContainsAny(kEventThreadNetdataChanged | kEventThreadRoleChanged))
```

Comment:
is it necessary to process for `kEventThreadRoleChanged` here? 
srp server itself processes role change event to publish or unpublish accordingly.
https://github.com/openthread/openthread/blob/ac21e00fa160988619c6a8a55967a92f436fb54b/src/core/net/srp_server.cpp#L492

Reply:
Yes. It is necessary.

`Publisher` can be used independent of whether OT SRP server feature is being used or not (e.g. in cases where app/next layer is providing SRP sever functionality).